slurmutils/editors/editor.py

```diff
@@ -0,0 +1,131 @@
+# Copyright 2024 Canonical Ltd.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Base methods for Slurm workload manager configuration file editors."""
+
+import logging
+import shlex
+from functools import wraps
+from os import path
+from typing import Any, Dict, List, Tuple
+
+from ..exceptions import EditorError
+
+_logger = logging.getLogger("slurmutils")
+
+
+def _is_comment(line: str) -> bool:
+    """Check if line is a comment."""
+    return line.startswith("#")
+
+
+def _contains_comment(line: str) -> bool:
+    """Check if line contains an inline comment."""
+    return "#" in line
+
+
+def _slice_comment(line: str) -> str:
+    """Slice inline comment off of line."""
+    return line.split("#", maxsplit=1)[0]
+
+
+def clean(line: str) -> Tuple[str, bool]:
+    """Clean line before further processing.
+
+    Returns:
+        Returns the cleaned line and False if it should be ignored.
+        If True, then the processors should ignore the line.
```

Comment:
This distinguishes between lines that start with "#" and lines that start with whitespace, but also have a comment at the end (" # indented comment"), making it throw an error in the second case (probably). Maybe we should just use the empty string as our marker to indicate that the line should be ignored.

Reply:
I prefer to have the boolean flag to indicate whether the line should be processed or not. I wanted a similar implementation like Go maps where:

```go
v, ok = m["NodeName"]
if !ok {
        fmt.Println("Oops - forgot the node name")
       os.Exit(1)
}
```

That way we don't need to rely on a default value for the primary argument, and we can just choose to ignore the second return if we know that the output we want is good.

Reply:
Then `Optional[str]` should be enough in that case, returning `None` if the line shouldn't be processed. I don't think we need to rely on go-isms since Python natively supports optionals.

Reply:
_But but... MY GOISMS!! :exploding_head:_

Good compromise for me :+1:

Reply:
Less Go-ish in https://github.com/charmed-hpc/slurmutils/pull/15/commits/87e6fb2c690a86e6fccdd662acdf5806da72648c